docs/architecture.md

```diff
@@ -0,0 +1,173 @@
+
+### JSON Schema generation
+
+You may have noticed that the previous serialization core schema has a `return_schema` key.
+This is because the core schema is also used to generate the corresponding JSON Schema.
+
+The JSON Schema generation of a model is handled by the [`GenerateJsonSchema`][pydantic.json_schema.GenerateJsonSchema] class.
+The [`generate`][pydantic.json_schema.GenerateJsonSchema.generate] method is the main entry point and is given the core schema of that model.
+
+### Customizing the core schema and JSON schema
+
+!!! abstract "Usage Documentation"
+    [Custom types](concepts/types.md#custom-types)
+
+While the `GenerateSchema` and [`GenerateJsonSchema`][pydantic.json_schema.GenerateJsonSchema] classes handle
+the creation of the corresponding schemas, Pydantic offers a way to customize them in some cases, following a wrapper pattern.
+This customization is done through the `__get_pydantic_core_schema__` and `__get_pydantic_json_schema__` methods.
+
+To understand this wrapper pattern, we will take the example of metadata classes used with [`Annotated`][typing.Annotated]
+where the `__get_pydantic_core_schema__` method can be used:
+
+```python
+from typing import Annotated, Any
+
+from pydantic import GetCoreSchemaHandler, TypeAdapter
+from pydantic_core import CoreSchema
+
+class MyStrict:
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+        schema = handler(source) # (1)!
+        schema['strict'] = True
+        return schema
+
+class MyGt:
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source: Any, handler: GetCoreSchemaHandler) -> CoreSchema:
+        schema = handler(source) # (2)!
+        schema['gt'] = 1
+        return schema
+
+ta = TypeAdapter(Annotated[int, MyStrict(), MyGt()])
+```
+
+1. `MyStrict` is the first annotation to be applied. At this point, `schema = {'type': 'int'}`.
+2. `MyGt` is the last annotation to be applied. At this point, `schema = {'type': 'int', 'strict': True}`.
+
+When the `GenerateSchema` class builds the core schema for `Annotated[int, MyStrict(), MyGt()]`, it will
+create an instance of a `GetCoreSchemaHandler` to be passed to the `MyGt.__get_pydantic_core_schema__` method. (1)
+{ .annotate }
+
+1.  In the case of our [`Annotated`][typing.Annotated] pattern, the `GetCoreSchemaHandler` is defined in a nested way.
+    Calling it will recursively call the other `__get_pydantic_core_schema__` methods until it reaches the `int` annotation,
+    where a simple `{'type': 'int'}` schema is returned.
+
+The `source` argument depends on the core schema generation pattern. In the case of [`Annotated`][typing.Annotated],
+the `source` will be the type being annotated. When [defining a custom type](concepts/types.md#as-a-method-on-a-custom-type),
+the `source` will be the actual class where `__get_pydantic_core_schema__` is defined.
+
+## Model validation and serialization
```

Comment:
This section should definitely be expanded to document _how_ this validation and serialization is performed (and not only state that it happens in `pydantic-core`). But I'll need to dive deeper in the Rust code base first.